geopmctlpy/geopmctl/metric.py

```diff
@@ -0,0 +1,680 @@
+        '''
+        Getter method for Signal report formatter.
+        '''
+        return self._report_formatter
+
+    def get_trace_formatter(self):
+        '''
+        Getter method for Signal trace formatter.
+        '''
+        return self._trace_formatter
+
+    def value(self):
+        '''
+        Returns
+
+        Computed value for this signal according to the last update value. None, if there is not
+        enough information (for example, update() is not called yet).
+        '''
+        return self._value
+
+    def reset(self):
+        '''
+        Return to a state where update() has not been called yet.
+        '''
+        self._value = None
+
+    def update(self, value):
+        '''
+        Update the value of the signal from a sample.
+
+        Args
+
+        value (float): Value returned by pio.
+        '''
+        self._value = value
+
+    def get_signature(self):
+        '''
+        Returns
+
+        A string of name@domain-instance.
+        '''
+        return \
+            f'{self._descriptor[0]}@{topo.domain_name(self._descriptor[1])}-{self._descriptor[2]}'
+
+
+class MonotoneSignal(Signal):
+    '''
+    Implementation of Signal where the computed value is delta between the last sample and current
+    sample for monotone signals.
+    '''
+
+    def __init__(self, descriptor, report_formatter=None, trace_formatter=None):
+        self._last_value = None
+        super().__init__(descriptor, Behavior.MONOTONE, report_formatter=report_formatter,
+                         trace_formatter=trace_formatter)
+
+    def reset(self):
+        self._last_value = None
+        super().reset()
+
+    def update(self, value):
+        if self._last_value is None:
+            self._last_value = value
+        else:
+            self._value = value - self._last_value
+            self._last_value = value
+
+
+class Algorithm(ABC):
+    '''
+    A class that represents an algorithm that runs in an Agent. the inputs to an Algorithm are
+    passed via a MetricList object. Algorithm processes the metrics and returns a list of floats
+    for control values.
+    '''
+
+    def __init__(self, metric_list):
+        '''
+        Args
+
+        metric_list (MetricList): A MetricList object that represents the input values to the
+        algorithm.
+        '''
+        self._metric_list = metric_list
+        self._controls = None
+
+    @abstractmethod
+    def update(self):
+        '''
+        Update the algorithm state with the latest metric values. Should be called after call to
+        MetricList.update()
+        '''
+
+    def get_metric_list(self):
+        '''
+        Returns
+
+        The MetricList object stored in this instance.
+        '''
+        return self._metric_list
+
+    def get_controls(self):
+        '''
+        Returns
+
+        List of float: The controls that should be returned by Agent.update().
+        '''
+        return self._controls
+
+
+class Metric:
```

Comment:
This class has a lot of responsibilities.  I think it could be broken up a bit, and made to align better with the C++ code that's similar.

First, we've got the notion of Signals that are derived from multiple signals in 2 different forms in the C++ code: composite signals and combined signals.  Here are some examples:
```
service/src/DerivativeSignal.hpp
service/src/DifferenceSignal.hpp
service/src/MultiplicationSignal.hpp
```
These composite signals derive from Signal and perform the necessary conversion for their type (e.g. difference or multiplication) in their ```read()``` and ```sample()``` methods.  I think you can do something similar here and avoid putting all this logic into one class.

Combined signals (derived from CombinedSignal.cpp) can probably be ignored for the moment as they aren't being used.  They provide the ability to actually combine 2 named signals into a new named signal.  This may or may not be useful here.

On the topic of aggregation/accumulation, in the C++ code this is handled by the SampleAggregator and ProcessRegionAggregator.  I realize that the Python runtime is a simpler implementation of what the C++ Controller does, so you can probably just focus on the SampleAggregator.  The SampleAggregator maintains the state of accumulated value that will ultimately be written to the Report.

It seems like a separate class should be created here with the responsibility of maintaining the aggregated/accumulated state.  Further the Reporting and Tracing should also be broken out into their own classes.  I think you, Chris, and I might want to get together and talk about the approach here.

In general, this looks pretty different than how we handle things in the C++ code.  I will be difficult to maintian.

Reply:
I also think we should avoid the word "metric" if we really mean CompositeSignals and lists of signals.  It's confusing.